### PR TITLE
Add eMMC plugin

### PR DESCRIPTION
**Pull request by superm1**
Not tested yet (no binaries available)

Type of pull request:
- [x] New plugin (Please include [new plugin checklist](https://github.com/hughsie/fwupd/wiki/New-plugin-checklist))

- [x] Fill out README.md with update protocol
- [x] Fill out README.md with any custom quirks and flags
- [x] CI run for at least one target
- [x] Document targets CI isn't currently run and the reasons (i.e. minimum versions needed or distribution limitations etc).

### CODE REVIEW

**Comment by hughsie** [comment]
I've pushed a fixup, I hope that's okay.

File: plugins/emmc/mmc.h [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 021110-1307, USA.
+ *
+ * Modified to add field firmware update support,
+ * those modifications are Copyright (c) 2016 SanDisk Corp.
+ */
```
Comment (hughsie) [inline plugins/emmc/mmc.h]: Can we source this header from the kernel somehow?

File: src/fu-engine.c [hunk 2]
```diff
@@ -3958,26 +3958,31 @@ fu_engine_udev_device_add (FuEngine *self, GUdevDevice *udev_device)
...
-			g_warning ("failed to find specified plugin %s: %s",
-				   plugin_name, error->message);
-			return;
-		}
-		if (!fu_plugin_runner_udev_device_added (plugin, device, &error)) {
-			if (g_error_matches (error, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED)) {
-				g_debug ("%s ignoring: %s",
-					 fu_plugin_get_name (plugin),
-					 error->message);
+		g_auto(GStrv) plugins = g_strsplit (plugin_name, ",", -1);
```
Comment (hughsie) [inline src/fu-engine.c]: this is correct here, but won't be supported by the libxmlb version. I'll have a think on how we can handle this kind of case better.

File: plugins/emmc/mmc.h (hunk 1 above)
Comment (superm1) [reply plugins/emmc/mmc.h]: It comes from the mmc-utils userspace package

**Comment by superm1** [comment]
Thanks for the fix-up!

File: plugins/emmc/mmc.h (hunk 1 above)
Comment (hughsie) [reply plugins/emmc/mmc.h]: That doesn't export a devel header, does it?

Comment (superm1) [reply plugins/emmc/mmc.h]: No, it doesn't.  So the options are either carry it verbatim or just pull out the `#define`s needed.

Comment (hughsie) [reply plugins/emmc/mmc.h]: unless we need more than about 50% of those, I'm erring on just pulling the ones we need.

Comment (superm1) [reply plugins/emmc/mmc.h]: Seems that we need about 20%, I pulled them in.

**Comment by superm1** [comment]
I don't really understand the Fedora CI failure; specifically how that's happening from changes to a different plugin.  Can you repro it?

**Comment by hughsie** [comment]
I guess Fedora is shipping a new libsmbios which is leaking memory, and thus libasan gets upset.

**Comment by superm1** [comment]
> I guess Fedora is shipping a new libsmbios which is leaking memory, and thus libasan gets upset.

There hasn't been a new libsmbios release though, and why is it only happening in this PR?

**Comment by superm1** [comment]
Now I see it happening in some other ones too, so that is reassuring at least.

**Comment by superm1** [comment]
Looking at https://koji.fedoraproject.org/koji/buildinfo?buildID=1394098 it doesn't seem to have any changes though lately other than a rebuild.

File: plugins/emmc/fu-emmc-device.c [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+	if (ext_csd[EXT_CSD_REV] < EXT_CSD_REV_V5_0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "FFU is only available on devices >= "
+			     "MMC 5.0, not supported in %s",
+			     fu_device_get_name (device));
+		return FALSE;
+	}
+	if (!(ext_csd[EXT_CSD_SUPPORTED_MODES] & EXT_CSD_FFU)) {
```
Comment (hughsie) [inline plugins/emmc/fu-emmc-device.c]: == 0?

File: plugins/emmc/fu-emmc-device.c [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+fu_emmc_device_prepare_firmware (FuDevice *device,
+				 GBytes *fw,
+				 FwupdInstallFlags flags,
+				 GError **error)
+{
+	FuEmmcDevice *self = FU_EMMC_DEVICE (device);
+	gsize fw_size = g_bytes_get_size (fw);
+
+	/* check alignment */
+	if (fw_size % self->sect_size) {
```
Comment (hughsie) [inline plugins/emmc/fu-emmc-device.c]: Maybe > 0 might be clearer?

**Comment by superm1** [comment]
OK with the branch as is, everything enumerates on the chromebook w/ eMMC I have on hand like so:

```
FuEmmcDevice:
DF4032
  DeviceId:             e732ac13d9601c6a7ada7559ee456be0e76308fa
  Guid:                 a884e2a1-d986-53d9-bbc2-277081d64fce <- EMMC\DF4032
  Guid:                 7d40e168-5054-589d-8446-0a50ed724ea5 <- EMMC\0045&0100
  Guid:                 a3e9470f-da57-5b0c-8bbd-95c2f305cec5 <- EMMC\0045&0100&DF4032
  Serial:               0x32e47300
  Plugin:               emmc
  Flags:                internal|updatable|registered
  VendorId:             0x000045
  Version:              0x3230343037332020
  VersionFormat:        number
  Icon:                 media-memory
  Created:              2019-10-18
  PhysicalId:            DEVPATH=/devices/pci0000:00/0000:00:1c.0/mmc_host/mmc1/mmc1:0001
  FD:                    14
  SectorSize:            512
```

Vendor ID 0x45 isn't listed in mmc-utils source code, so I don't know who actually makes this disk and don't have binaries to try to flash.

**Comment by superm1** [comment]
I'll clean up those two nits.  Any thoughts on the protocol I selected?  The alternative I was thinking was `org.jedec.mmc`

**Comment by superm1** [comment]
> org.jedec

They have a bunch of standards, and this is super MMC specific, so i'm switching it to `org.jedec.mmc`

**Comment by hughsie** [comment]
> They have a bunch of standards, and this is super MMC specific, so i'm switching it to org.jedec.mmc

If it helps, this is exactly what I was going to  say :)

**Pull request merged**